### PR TITLE
ET-4605 and or filter

### PR DESCRIPTION
**Pull request by janpetschexain**
**Reference**

- [ET-4605]
- requires #1003
- followed by #1016

**Summary**

- declare and/or filters in the spec & update changelog
- impl combine filter serde
- add combine filter in elasticsearch calls

[ET-4605]: https://xainag.atlassian.net/browse/ET-4605?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: web-api/src/storage/elastic/filter.rs [hunk 1]
```diff
@@ -12,36 +12,81 @@
...
+                *filter = json!([filter.take(), clause]);
+            }
+            _ => unreachable!(/* filter and clause are array or object */),
+        }
+    } else {
+        filter.insert(occurence.to_string(), clause);
+    }
+}
+
+fn extend_filter(filter: &mut JsonObject, clause: &Filter, is_not_root: Option<&'static str>) {
```
Comment (rustonaut) [inline web-api/src/storage/elastic/filter.rs:43]: I will approve it for now but I feel the readability and maintainability of this code is pretty bad.

Some ideas:

- maybe we probably should not special case "is root"/"is not root"
- avoid or change how we do some of the micro optimizations like the `"should": [{...}]` as `"should": {...}` part

File: web-api/tests/filter.rs [hunk 2]
```diff
@@ -391,3 +391,162 @@ fn test_filter_array_string_multiple() {
...
+            let documents = send_assert_json::<SemanticSearchResponse>(
+                &client,
+                client
+                    .post(personalization_url.join("/semantic_search")?)
+                    .json(&json!({ "document": { "query": "zero" }, "filter": { "$or": [] } }))
+                    .build()?,
+                StatusCode::OK,
+            )
+            .await;
+            assert_eq!(documents.ids(), ["d1", "d2", "d3"].into());
```
Comment (rustonaut) [inline web-api/tests/filter.rs:513]: :wrench:  There is a bug somewhere here I think.

`"$or": []` should always be false, i.e. not match any documents

I mean it should map to `{ "should":  [], "minimum_should_match": 1 }` which can never evaluate to have enough matches :thinking: 

(Also logically `$and: [ (...),*]` is like `true (&& ...)*` and `$or: [...,*]` like `false (|| ...)*` )

File: web-api/src/personalization/filter.rs [hunk 3]
```diff
@@ -147,10 +148,142 @@ impl<'de> Deserialize<'de> for Compare {
...
+                    } else {
+                        return Err(A::Error::invalid_length(filters.len() + 1, &Self));
+                    }
+                }
+
+                Ok(Filters(filters))
+            }
+        }
+
+        deserializer.deserialize_seq(FiltersVisitor)
```
[the remark this answers is not in the document]
Comment (janpetschexain) [reply web-api/src/personalization/filter.rs:211]: but this opens up for performance issues. if a request contains a very large filter (eg as much as the byte limit of the request allows) and we deserialize it completely and only afterwards check the length and discard it, then a lot of work was wasted.
the overhead for a custom visitor on the other hand is not too much as it is only a single additional function.

File: web-api/src/storage/elastic/filter.rs (hunk 1 above)
Comment (janpetschexain) [reply web-api/src/storage/elastic/filter.rs:43]: > maybe we probably should not special case "is root"/"is not root"

i agree with that, but it requires further refactoring in the es modules, because we would also need to apply the `{"bool": {"filter": object}}` structure to the root level, but currently we carry only the `object` thingy around. hence i didn't make these changes yet, but we definitely should follow up on this.

> avoid or change how we do some of the micro optimizations like the "should": [{...}] as "should": {...} part

i was wondering how es handles that internally, does it make a difference performance wise? or is it handled as array of objects anyways?

File: web-api/tests/filter.rs (hunk 2 above)
Comment (janpetschexain) [reply web-api/tests/filter.rs:513]: this is forwarded as is to es, so we probably have to decide if es's edge case handling is what we want or not.

File: web-api/src/personalization/filter.rs [hunk 4]
```diff
@@ -147,10 +148,142 @@ impl<'de> Deserialize<'de> for Compare {
...
+
+        deserializer.deserialize_map(CombineVisitor)
+    }
+}
+
 #[derive(Clone, Debug, Deserialize, PartialEq)]
 #[serde(untagged)]
 pub(crate) enum Filter {
     Compare(Compare),
+    Combine(Combine),
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply web-api/src/personalization/filter.rs:277]: it's not about performance but about making sure we never parse `$and` or `$or` accidentally as field names.

Like when adding `$document.tags` or similar .

File: web-api/src/personalization/filter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rustonaut) [reply web-api/src/personalization/filter.rs:211]: For preventing performance issue we have size limits on the JSON body, and wrt. the generated code it's much much more then just one function.

Through we can keep it that way, still I think we should in general prefer writing the simpler code and only do such optimizations if there are performance issues.

File: web-api/tests/filter.rs (hunk 2 above)
Comment (rustonaut) [reply web-api/tests/filter.rs:513]: ES :face_exhaling: 

I guess that becomes a [:pushpin: issue](https://xainag.atlassian.net/browse/ET-4738).

File: web-api/src/personalization/filter.rs [hunk 5]
```diff
@@ -147,10 +148,142 @@ impl<'de> Deserialize<'de> for Compare {
...
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct Combine {
+    pub(crate) operation: CombineOp,
+    pub(crate) filters: Filters,
+}
+
+impl Combine {
+    const MAX_LEVEL: usize = 2;
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply web-api/src/personalization/filter.rs:222]: you didn't push any changes

File: web-api/src/personalization/filter.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (janpetschexain) [reply web-api/src/personalization/filter.rs:277]: from a user input this can never happen as we don't allow `$` as a char in ids. if we on our side decide to add some special ids using forbidden chars, then we need to very carefully evaluate if and what to use, we probably need a list of reserved special ids then. until then we can keep it as is.

**Review by rustonaut — APPROVED** [review]
it would be nice to still change the `Combine::MAX_LEVEL` to a more clear name like `MAX_DEPTH`.

File: web-api/src/storage/elastic/filter.rs (hunk 1 above)
Comment (janpetschexain) [reply web-api/src/storage/elastic/filter.rs:41]: i simplified the code such that all clauses must be arrays now, this removes a lot of the array/object handling.

**Review by rustonaut — APPROVED** [review]
